### ISSUE TITLE
JSON import plugin is broken (infinite loop in production build, blank page in dev mode)

### ISSUE THREAD

**Issue by danielweck** [preactjs/wmr#446]
**Describe the bug**

* `npm run build` => shell command hangs forever, and the following addition to `wmr.config.mjs` reveals the infinite loop:

```js
function tracePlugin() {
    return {
        name: 'trace',
        async resolveId(id, _importer) {
            console.log('##### resolveId ', id);
            return undefined;
        },
    };
}
config.plugins.push(tracePlugin);
```

* `npm start` => the page remains blank

**To Reproduce**

1) `npm init wmr json-import`
2) `cd json-import`
3) open `public/pages/_404.js`, add at the top: `import test from './test.json';` or `import test from 'json:./test.json';` (both trigger the bug).
4) add `<pre>{json}</pre>` in the JSX
5) create `test.json` in `public/pages/`, for example (**)
4) `npm run build` or `npm start` + open web browser URL

(**) `test.json`
=>
```json
{
    "key": "value",
    "obj": {
        "array": [1, 2, 3]
    },
    "boolean": false,
    "null": null
}
```

Code culprit (`this.resolve()` seems to call `resolveId()`, recursively):

https://github.com/preactjs/wmr/blob/7e1fb016be6bf91eeaa424a452b1bb215629a52e/packages/wmr/src/plugins/json-plugin.js#L18-L28

**Expected behavior**

The JSON import should work.

**Desktop (please complete the following information):**
 - OS: all
 - Browser: all
 - WMR Version: all

**Additional context**

**Comment by developit** [comment]
Hi @danielweck - want me to land a fix for this, or are you putting one together?

**Comment by danielweck** [comment]
I have some sort of local fix, but I am working with the official WMR release (using my own userland plugin) so I have to circumvent WMR's built-in JSON plugin by using a different import prefix (`jsonx:`) and a different file extension (`.jsonx`), which I suspect isn't representative of a real-world fix because if I understand correctly the `.json` file extension is getting some special treatment in WMR (like `.js` etc.) whereas my ad-hoc `.jsonx` file extension is completely unrecognized by WMR.

Long story short: I would really appreciate if you could fix this at your end properly :)
Many thanks!

**Comment by danielweck** [comment]
That's my local fix:
```js

	function jsonXplugin() {
		const IMPORT_PREFIX = 'jsonx:';
		const INTERNAL_PREFIX = '\0jsonx:';
		return {
			name: 'jsonx-plugin',
			async resolveId(id, importer) {
				if (id[0] === '\0' || id[0] === '\b') return;

				if (id.startsWith(IMPORT_PREFIX)) {
					id = id.slice(IMPORT_PREFIX.length);
				} else if (config.mode === 'build') {
					return;
				} else if (!id.endsWith('.jsonx')) {
					return;
				}

				const resolved = await this.resolve(id, importer, { skipSelf: true });
				if (!resolved) return;

				resolved.id = `${INTERNAL_PREFIX}${resolved.id}`;
				return resolved;
			},
			async load(id) {
				if (!id.startsWith(INTERNAL_PREFIX)) return;

				id = id.slice(INTERNAL_PREFIX.length);

				id = path.resolve(config.cwd || '.', id);
				this.addWatchFile(id);
				const j = fs.readFileSync(id, { encoding: 'utf8' });
				return `export default ${j}`;
			},
		};
	}
	config.plugins.push(jsonXplugin());
```

**Comment by danielweck** [comment]
Ah sorry, I forgot the non-prefixed import case. I corrected the above code snippet. I tested these cases:

1) `npm start` (`config.mode === 'start'`)
2) `npm run build --prerender && npm run serve` (`config.mode === 'build'`)

Each with the prefixed and un-prefixed import:

* `import json from 'jsonx:./test.jsonx';`
* `import json from './test.jsonx';`

**Comment by danielweck** [comment]
As suspected, If I change the `.jsonx` file extension to `.json` in my tests (whilst keeping the `jsonx` custom import prefix in my userland plugin), then my `console.log()` shows `\x00json:\x00jsonx:/PATH/TO/public/pages/test.json, which indicates that WMR's built-in handler gets involved in the plugin chain.

**Comment by danielweck** [comment]
I am experiencing a similar issue with SVG files, and my workaround is based on the same technique used to bypass this JSON bug: https://github.com/preactjs/wmr/issues/449

**Comment by danielweck** [comment]
So, I am running some tests from the `main` branch which includes the fix https://github.com/preactjs/wmr/pull/480
...however I am now experiencing pretty serious regressions (router, and `text:` plugin). I am investigating...

**Comment by marvinhagemeister** [comment]
@danielweck do you have a repo to look at? Happy to help debugging it

**Comment by danielweck** [comment]
I think this is a false alert: most likely a cache busting issue ... continuing tests :)

**Comment by danielweck** [comment]
Apologies for raising a false alarm, I had a HTTP server / cache issue.
Thank you so much for solving the JSON bug during your weekend hours 👍  :)
